### ISSUE TITLE
R1732(consider-using-with): Regression on `contextlib.ExitStack`

### ISSUE THREAD

**Issue by fmigneault** [pylint-dev/pylint#9162]
### Bug description

```python
This is extremely similar to https://github.com/pylint-dev/pylint/issues/4654, but is somehow still flagged in my case. 

I can see 2 possible distinctions. 

1. I use  `stack.enter_context(tempfile.TemporaryDirectory())` and `stack.enter_context(tempfile.NamedTemporaryFile(...))` instead of `stack.enter_context(open(...))` as in the original issue.

2. The `with contextlib.ExitStack` block is not directly in the same context/scope as where `stack.enter_context` is called. 

For example:


def util_func(stack):
    tmpdir = stack.enter_context(tempfile.TemporaryDirectory())
    # do stuff
    return tmpdir

def test_something():
    with contextlib.ExitStack() as stack:
        tmpdir = util_func(stack)
        stack.enter_context(tempfile.NamedTemporaryFile(dir=tmpdir))
        # ...
```

### Command used

```shell
pylint \
	--load-plugins pylint_quotes \
	--rcfile=./.pylintrc \
	--reports y \
    ./weaver ./tests
```

### Pylint output

```python
************* Module tests.functional.test_builtin
tests/functional/test_builtin.py:93:38: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
tests/functional/test_builtin.py:401:38: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
tests/functional/test_builtin.py:403:12: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
```

### Expected behavior

No error for `stack.enter_context(tempfile.TemporaryDirectory())` and `stack.enter_context(tempfile.NamedTemporaryFile(...))`. 

### Pylint version

```shell
Name: pylint
Version: 2.17.4
Summary: python code static checker
Home-page: 
Author: 
Author-email: Python Code Quality Authority <code-quality@python.org>
License: GPL-2.0-or-later
Location: /home/francis/dev/miniconda/envs/weaver-py310/lib/python3.10/site-packages
Requires: astroid, dill, isort, mccabe, platformdirs, tomli, tomlkit
Required-by: pylint-quotes
```

### OS / Environment

Ubuntu 20.04.6
